checkov/azure_pipelines/image_referencer/provider.py

```diff
@@ -0,0 +1,56 @@
+from __future__ import annotations
+
+from typing import Any
+
+from checkov.common.images.image_referencer import Image
+from checkov.azure_pipelines.common.resource_id_utils import generate_resource_key_recursive
+from checkov.common.images.image_referencer_provider import Provider
+
+
+class AzurePipelinesProvider(Provider):
+
+    def __init__(self, workflow_config: dict[str, Any], file_path: str):
+        super().__init__(workflow_config, file_path)
+        self.supported_keys = "container"
+
+    def extract_images_from_workflow(self) -> list[Image]:
+        images = self.extract_images_from_dict(self.workflow_config)
+        return images
+
+    def extract_images_from_list(self, objects_list: list[dict[str, Any]]) -> list[Image]:
+        images = []
+        for job in objects_list:
+            if isinstance(job, dict):
+                images.extend(self.extract_images_from_dict(job))
+            if isinstance(job, list):
+                images.extend(self.extract_images_from_list(job))
+        return images
+
+    def extract_images_from_dict(self, job: dict[str, Any]) -> list[Image]:
+        images = []
+        start_line, end_line = AzurePipelinesProvider._get_start_end_lines(job)
+        for key, sub_job in job.items():
+            if key == self.supported_keys:
+                images.append(self.crete_image(sub_job, start_line, end_line))
+            elif isinstance(sub_job, dict):
+                images.extend(self.extract_images_from_dict(sub_job))
+            elif isinstance(sub_job, list):
+                images.extend(self.extract_images_from_list(sub_job))
+        return images
```

Comment:
the biggest drawback of this approach is that we basically do a deep search for the key `container`. Depending on the size of the file, this can be rather time consuming. Wouldn't it be much faster just to directly search under the keys, where we expect a `container` reference?
ex
```python
for job in self.workflow_config.get("jobs"):
    container = job.get("container")
    if container:
        images.append(self.create_image(container, start_line, end_line))
```
and for `stages` something similar.

Reply:
I think all the relevant paths to an image are listed here https://learn.microsoft.com/en-us/azure/devops/pipelines/process/container-phases?view=azure-devops

You can also use jmespath to traverse the json with keywords - example in `checkov/circleci_pipelines/image_referencer/provider.py`

Reply:
the problem is that i can have 
     'stage'-> 'job'->'steps'->'task'->'container' 
or  'job' -> 'container' 
or 'job' -> 'container' ->'image'
and not directly in 'job' or in 'stage'.
@gruebel

Reply:
the problem is that i can have
'stage'-> 'job'->'steps'->'task'->'container'
or 'job' -> 'container'
or 'job' -> 'container' ->'image'
and not directly in 'job' or in 'stage'.